### PR TITLE
Adds Native SSO Token Exchange Support (https://developer.okta.com/docs/guides/configure-native-sso/main/)

### PR DESCRIPTION
**Pull request by emanor-okta**
#### Description:
:seedling: Adds support for EA feature [Native SSO Token Exchange](https://developer.okta.com/docs/guides/configure-native-sso/main)

#### Testing details:
- [X]  Verified basic functionality of change
- [ ]  Added tests 

#### Other considerations:
- [X] Has security implications
- [ ] Has UX changes

##### RESOLVES: 
N/A

### CODE REVIEW

**Comment by JayNewstrom** [comment]
Hi, thanks for the PR. We will review and pull it in if we can. In the meantime, can you sign our CLA? 

https://developer.okta.com/cla/

Internal Ref: OKTA-422248

**Review by JayNewstrom — COMMENTED** [review]
I think this is something we definitely want to bring in. Could you add some tests? It also looks like checkstyle is complaining, and you're missing a few headers for license. 

If you're not willing to take this through to the finish line, let me know and I'll get it prioritize on our team.

File: library/src/main/java/com/okta/oidc/clients/BaseAuth.java [hunk 1]
```diff
@@ -30,6 +30,10 @@
      * When set, signing out will attempt to revoke the refresh token.
      */
     int REVOKE_REFRESH_TOKEN = 0x00000002;
+    /**
+     * When set, signing out will attempt to revoke the refresh token.
```
Comment (NikitaAvraimov-okta) [inline library/src/main/java/com/okta/oidc/clients/BaseAuth.java:34]: revoke device secret?

File: library/src/main/java/com/okta/oidc/net/request/HttpRequestBuilder.java [hunk 2]
```diff
@@ -218,6 +222,45 @@ public TokenRequest createRequest() throws AuthorizationException {
...
+
+        public DeviceSecretTokenExchange subjectToken(String subjectToken) {
+            mSubjectToken = subjectToken;
+            return this;
+        }
+
+        @Override
+        protected void validate(boolean isConfigurationRequest) throws AuthorizationException {
+            super.validate(isConfigurationRequest);
+            if (mConfig.getAudience() == null || mConfig.getAudience().equalsIgnoreCase("")) {
```
Comment (NikitaAvraimov-okta) [inline library/src/main/java/com/okta/oidc/net/request/HttpRequestBuilder.java:251]: isEmpty() looks kind of better here (nit)

File: library/src/main/java/com/okta/oidc/net/request/DeviceSecretTokenRequest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        params.put("client_id", client_id);
+        params.put("grant_type", grant_type);
+        params.put("actor_token", b.mActorToken);
+        params.put("actor_token_type", TokenTypeHint.ACTOR_TOKEN_TYPE);
+        params.put("subject_token", b.mSubjectToken);
+        params.put("subject_token_type", TokenTypeHint.SUBJECT_TOKEN_TYPE);
+        //requires API 24
+        //params.put("scope", Arrays.stream(b.mConfig.getScopes()).reduce("", (a, e) -> a + " " + e));
+        String scope = b.mConfig.getScopes()[0];
+        for (int i=1; i<b.mConfig.getScopes().length; scope += " " + b.mConfig.getScopes()[i++])
```
Comment (NikitaAvraimov-okta) [inline library/src/main/java/com/okta/oidc/net/request/DeviceSecretTokenRequest.java:51]: TextUtils.join(" ", b.mConfig.getScopes()) could be more clear in this case

File: library/src/main/java/com/okta/oidc/clients/BaseAuth.java (hunk 1 above)
Comment (emanor-okta) [reply library/src/main/java/com/okta/oidc/clients/BaseAuth.java]: fixed https://github.com/okta/okta-oidc-android/pull/262/commits/a5c97e4bedf174370a98369157c0f032ee193431

File: library/src/main/java/com/okta/oidc/net/request/HttpRequestBuilder.java (hunk 2 above)
Comment (emanor-okta) [reply library/src/main/java/com/okta/oidc/net/request/HttpRequestBuilder.java]: fixed https://github.com/okta/okta-oidc-android/pull/262/commits/a5c97e4bedf174370a98369157c0f032ee193431

**Comment by JayNewstrom** [comment]
For anyone interesting in trying out this functionality, we just released the first public beta of a new suite of SDKs here:
- Auth Foundation - https://search.maven.org/artifact/com.okta.kotlin/auth-foundation/0.2.0-BETA/aar
- Web Authentication UI - https://search.maven.org/artifact/com.okta.kotlin/web-authentication-ui/0.2.0-BETA/aar
- OAuth2 - https://search.maven.org/artifact/com.okta.kotlin/oauth2/0.2.0-BETA/aar
- IDX Kotlin - https://search.maven.org/artifact/com.okta.android/okta-idx-kotlin/2.0.0-BETA2/aar

With repos available here:
- https://github.com/okta/okta-mobile-kotlin/releases/tag/0.2.0-BETA
- https://github.com/okta/okta-idx-android/releases/tag/2.0.0-BETA2

The new functionality being focused around `TokenExchangeFlow`: https://github.com/okta/okta-mobile-kotlin/blob/master/oauth2/src/main/java/com/okta/oauth2/TokenExchangeFlow.kt

We welcome feedback (notes on how to submit feedback in the projects README)

**Comment by JayNewstrom** [comment]
No plans to merge. Please see the [new SDK](https://github.com/okta/okta-mobile-kotlin).

**Pull request closed without merging**